File: packages/azure_pronunciation_assessment/ios/Classes/AzurePronunciationAssessmentPlugin.swift
import Flutter
import MicrosoftCognitiveServicesSpeech

public class AzurePronunciationAssessmentPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: "azure_pronunciation_assessment",
      binaryMessenger: registrar.messenger())
    channel.setMethodCallHandler { call, result in
      guard call.method == "assess" else {
        result(FlutterMethodNotImplemented)
        return
      }
      guard let args = call.arguments as? [String: Any] else {
        result(FlutterError(code: "bad_args", message: "Expected map", details: nil))
        return
      }
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          let json = try Self.performAssessment(args: args)
          DispatchQueue.main.async { result(json) }
        } catch let e as NSError where e.domain == "AzurePronunciationAssessment" {
          let code = e.code == 1 ? "no_speech" : "azure_speech_error"
          DispatchQueue.main.async {
            result(FlutterError(code: code, message: e.localizedDescription, details: nil))
          }
        } catch {
          DispatchQueue.main.async {
            result(FlutterError(code: "azure_speech_error", message: error.localizedDescription, details: nil))
          }
        }
      }
    }
  }

  private static func performAssessment(args: [String: Any]) throws -> String {
    let audioPath = args["audioPath"] as! String
    let referenceText = args["referenceText"] as! String
    let language = args["language"] as! String
    let token = args["token"] as! String
    let region = args["region"] as! String
    let enableProsody = (args["enableProsody"] as? Bool) ?? true
    let enableMiscue = (args["enableMiscue"] as? Bool) ?? true
    let nbestPhonemeCount = (args["nbestPhonemeCount"] as? Int) ?? 1
    let phonemeAlphabet = (args["phonemeAlphabet"] as? String) ?? "IPA"
    let granularityStr = (args["granularity"] as? String) ?? "Phoneme"

    let granularity: SPXPronunciationAssessmentGranularity
    switch granularityStr {
    case "Word": granularity = .word
    case "FullText": granularity = .fullText
    default: granularity = .phoneme
    }

    let speechConfig = try SPXSpeechConfiguration(authorizationToken: token, region: region)
    try speechConfig.setSpeechRecognitionLanguage(language)

    guard let audioConfig = SPXAudioConfiguration(wavFileInput: audioPath) else {
      throw NSError(
        domain: "AzurePronunciationAssessment", code: 4,
        userInfo: [NSLocalizedDescriptionKey: "Could not open audio file"])
    }

    let pronunciationConfig = try SPXPronunciationAssessmentConfiguration(
      referenceText,
      gradingSystem: .hundredMark,
      granularity: granularity,
      enableMiscue: enableMiscue)
    if enableProsody {
      pronunciationConfig.enableProsodyAssessment()
    }
    pronunciationConfig.phonemeAlphabet = phonemeAlphabet
    pronunciationConfig.nbestPhonemeCount = UInt(nbestPhonemeCount)

    guard
      let speechRecognizer = try? SPXSpeechRecognizer(
        speechConfiguration: speechConfig, audioConfiguration: audioConfig)
    else {
      throw NSError(
        domain: "AzurePronunciationAssessment", code: 5,
        userInfo: [NSLocalizedDescriptionKey: "Could not create speech recognizer"])
    }

    try pronunciationConfig.apply(to: speechRecognizer)

    let semaphore = DispatchSemaphore(value: 0)
    var jsonOut: String?
    var errOut: NSError?

    speechRecognizer.recognizeOnceAsync { result in
      defer { result.close() }
      if result.reason == SPXResultReason.recognizedSpeech {
        jsonOut = result.properties?.getPropertyBy(SPXPropertyId.speechServiceResponseJsonResult)
        if jsonOut == nil || jsonOut!.isEmpty {
          errOut = NSError(
            domain: "AzurePronunciationAssessment", code: 2,
            userInfo: [NSLocalizedDescriptionKey: "Empty JsonResult"])
        }
      } else if result.reason == SPXResultReason.noMatch {
        errOut = NSError(
          domain: "AzurePronunciationAssessment", code: 1,
          userInfo: [NSLocalizedDescriptionKey: "No speech detected"])
      } else if result.reason == SPXResultReason.canceled {
        let cancellationDetails = try? SPXCancellationDetails(fromCanceledRecognitionResult: result)
        let msg = "\(String(describing: cancellationDetails?.reason)): \(cancellationDetails?.errorDetails ?? "")"
        errOut = NSError(
          domain: "AzurePronunciationAssessment", code: 3,
          userInfo: [NSLocalizedDescriptionKey: msg])
      } else {
        errOut = NSError(
          domain: "AzurePronunciationAssessment", code: 3,
          userInfo: [NSLocalizedDescriptionKey: "Unexpected result reason"])
      }
      semaphore.signal()
    }

    semaphore.wait()

    speechRecognizer.close()
    if let e = errOut { throw e }
    guard let json = jsonOut, !json.isEmpty else {
      throw NSError(
        domain: "AzurePronunciationAssessment", code: 2,
        userInfo: [NSLocalizedDescriptionKey: "No assessment JSON"])
    }
    return json
  }
}

File: packages/azure_pronunciation_assessment/android/src/main/kotlin/dev/enjoy_player/azure_pronunciation_assessment/AzurePronunciationAssessmentPlugin.kt
package dev.enjoy_player.azure_pronunciation_assessment

import android.os.Handler
import android.os.Looper
import com.microsoft.cognitiveservices.speech.*
import com.microsoft.cognitiveservices.speech.audio.AudioConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.Executors

class AzurePronunciationAssessmentPlugin : FlutterPlugin, MethodCallHandler {
  private lateinit var channel: MethodChannel
  private val executor = Executors.newSingleThreadExecutor()
  private val mainHandler = Handler(Looper.getMainLooper())

  override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(binding.binaryMessenger, "azure_pronunciation_assessment")
    channel.setMethodCallHandler(this)
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    if (call.method != "assess") {
      result.notImplemented()
      return
    }
    @Suppress("UNCHECKED_CAST")
    val args = call.arguments as? Map<String, Any?>
    if (args == null) {
      result.error("bad_args", "Expected map arguments", null)
      return
    }
    executor.execute {
      try {
        val json = assess(args)
        mainHandler.post { result.success(json) }
      } catch (e: Exception) {
        val code = mapErrorCode(e)
        mainHandler.post {
          result.error(code, e.message ?: code, null)
        }
      }
    }
  }

  private fun assess(args: Map<String, Any?>): String {
    val audioPath = args["audioPath"] as String
    val referenceText = args["referenceText"] as String
    val language = args["language"] as String
    val token = args["token"] as String
    val region = args["region"] as String
    val enableProsody = args["enableProsody"] as? Boolean ?: true
    val enableMiscue = args["enableMiscue"] as? Boolean ?: true
    val nbestPhonemeCount = (args["nbestPhonemeCount"] as? Number)?.toInt() ?: 1
    val phonemeAlphabet = args["phonemeAlphabet"] as? String ?: "IPA"
    val granularityStr = args["granularity"] as? String ?: "Phoneme"
    val granularity = parseGranularity(granularityStr)

    val config = SpeechConfig.fromAuthorizationToken(token, region)
    try {
      config.speechRecognitionLanguage = language
      val audioConfig = AudioConfig.fromWavFileInput(audioPath)
      try {
        val pronunciationConfig = PronunciationAssessmentConfig(
          referenceText,
          PronunciationAssessmentGradingSystem.HundredMark,
          granularity,
          enableMiscue,
        )
        try {
          pronunciationConfig.enableProsodyAssessment = enableProsody
          pronunciationConfig.phonemeAlphabet = phonemeAlphabet
          pronunciationConfig.nbestPhonemeCount = nbestPhonemeCount

          val recognizer = SpeechRecognizer(config, audioConfig)
          try {
            pronunciationConfig.applyTo(recognizer)
            val speechResult = recognizer.recognizeOnceAsync().get()
            when (speechResult.reason) {
              ResultReason.RecognizedSpeech -> {
                val json =
                  speechResult.properties.getProperty(
                    PropertyId.SpeechServiceResponse_JsonResult,
                  )
                if (json.isNullOrBlank()) {
                  throw IllegalStateException("Azure returned empty JsonResult property")
                }
                return json
              }
              ResultReason.NoMatch -> {
                throw NoSpeechException()
              }
              else -> {
                val cancel = CancellationDetails.fromResult(speechResult)
                throw IllegalStateException(
                  "${cancel.reason}: ${cancel.errorDetails}",
                )
              }
            }
          } finally {
            recognizer.close()
          }
        } finally {
          pronunciationConfig.close()
        }
      } finally {
        audioConfig.close()
      }
    } finally {
      config.close()
    }
  }

  private fun parseGranularity(s: String): PronunciationAssessmentGranularity =
    when (s) {
      "Word" -> PronunciationAssessmentGranularity.Word
      "FullText" -> PronunciationAssessmentGranularity.FullText
      else -> PronunciationAssessmentGranularity.Phoneme
    }

  private fun mapErrorCode(e: Exception): String =
    when (e) {
      is NoSpeechException -> "no_speech"
      else -> "azure_speech_error"
    }

  private class NoSpeechException : RuntimeException("No speech detected in the audio.")
}
